File: F#/p003-Dot-Product.fs
(*
    Doing dot product with a layer of neurons and multiple inputs
    Associated YT NNFS tutorial: https://www.youtube.com/watch?v=tMrbN67U9d4
*)

open NumSharp
(*
    NumSharp is a NumPy port for .NET
    https://github.com/SciSharp/NumSharp
    https://www.nuget.org/packages/NumSharp/
*)

(* inputs and weights have to be mutable so we can get their address
   since they are 'top-level' values *)
let mutable inputs  = NDArray [|1.0; 2.0; 3.0; 2.5|]
let mutable weights = NDArray [|[|  0.2;   0.8; -0.5;  1.0|]
                                [|  0.5; -0.91; 0.26; -0.5|]
                                [|-0.26; -0.27; 0.17; 0.87|]|]

let biases = NDArray [|2.0; 3.0; 0.5|]

let output = np.dot(&weights, &inputs) + biases

System.Console.WriteLine output

File: VB.NET/p003-Dot-Product.vb
Imports Accord.Math

Module p003_Dot_Product
    Public Sub Main(args As String())
        Dim inputs() As Double = {1.0, 2.0, 3.0, 2.5}

        ' Single Neuron
        'Dim weights() As Double = New Double() {0.2, 0.8, -0.5, 1.0}
        'Dim bias As Double = 2.0
        'Dim output As Double = Matrix.Dot(weights, inputs) + bias
        'Console.WriteLine(output)

        ' Multiple Neurons
        Dim weights = New Double(2, 3) {{0.2, 0.8, -0.5, 1.0}, {0.5, -0.91, 0.26, -0.5}, {-0.26, -0.27, 0.17, 0.87}}
        Dim biases() As Double = {2.0, 3.0, 0.5, 1.0}
        Dim outputs() As Double = weights.Dot(inputs).Add(biases)
        Console.WriteLine($"[{String.Join(", ", outputs)}]")
    End Sub
End Module
